File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/ui/MoneyFormat.kt
package lk.mageride.passenger.ui

import lk.mageride.shared.data.models.Money

/**
 * `48000` → `Rs 480`; `48050` → `Rs 480.50`.
 *
 * **Money is integer minor units everywhere** (root CLAUDE.md) and nothing on the wire is ever a
 * decimal, so the only place rupees exist is here, one step before a `Text`. Doing it at the call
 * site would be the arithmetic repeated on every screen that shows a fare — which is most of them.
 *
 * **`Rs` is not translated.** It is the currency symbol printed in D2' §A, and three identical
 * values in the three `strings.xml` files is what `StringResourceTest` reads as a translation
 * nobody did — the same argument `LanguageNames` and `PhoneNumber` make for their constants.
 *
 * The driver app has its own `MoneyFormat` and this is deliberately a second copy rather than a
 * shared one: `:shared` is `commonMain` and has neither `java.util.Locale` nor `String.format`, and
 * the two apps' number groupings are free to diverge. Recorded in the C079 handoff.
 */
internal object MoneyFormat {

    /** The currency prefix D2' §A prints on every passenger-facing amount. */
    const val PREFIX: String = "Rs"

    /** What is drawn where an amount is not known yet — a quote in flight, or one that failed. */
    const val EMPTY: String = "—"

    fun rupees(money: Money): String = rupees(money.amountMinor)

    /** The same, for a bare `…Minor` field with no [Money] wrapper on the wire. */
    fun rupees(amountMinor: Long): String {
        val negative = amountMinor < 0
        val absolute = if (negative) -amountMinor else amountMinor
        val text = group(absolute / MINOR_UNITS) + centsOf(absolute % MINOR_UNITS)
        return if (negative) "$PREFIX -$text" else "$PREFIX $text"
    }

    /** Thousands separators, by hand — `commonMain` has no locale-aware formatter to borrow. */
    private fun group(whole: Long): String = whole.toString()
        .reversed()
        .chunked(GROUP_SIZE)
        .joinToString(",")
        .reversed()

    /** Whole rupees lose their `.00`; anything else keeps two digits. */
    private fun centsOf(cents: Long): String = if (cents == 0L) "" else ".${cents.toString().padStart(2, '0')}"

    private const val MINOR_UNITS = 100L
    private const val GROUP_SIZE = 3
}

File: apps/passenger-ios/PassengerApp/UI/MoneyFormat.swift
import Foundation
import MageRideShared

/// Rendering a rupee amount — `Rs 740`, `Rs 1,240`, `Rs 480.50`.
///
/// **`Money` deliberately does not format itself** (C012): rendering needs a locale and a symbol,
/// and no user-facing string may live in `:shared`. This is the app's half of that split, and it is
/// one type so that SCR-PI-009's tier price, SCR-PI-012's estimate and C098's receipt all group
/// their thousands the same way. The same file is `apps/driver-ios/DriverApp/UI/MoneyFormat.swift`
/// and `apps/passenger-android/.../ui/MoneyFormat.kt`; C079's handoff argues why it is not in
/// `:shared` — `commonMain` has neither a locale nor a `String.format`, so a shared version would be
/// hand-rolled anyway.
///
/// **Everything here is integer arithmetic.** A rupee amount is `Int64` minor units on the wire and
/// stays that way to the last character; `Double` is the one thing C012's money fence forbids, and a
/// formatter that divided by 100 to print would reintroduce it at the only place a passenger reads
/// the number.
///
/// The grouping is fixed to `en_US_POSIX` rather than the handset's, because the figure belongs to a
/// Sri Lankan rupee amount: a passenger whose phone groups by lakhs must still read the number the
/// receipt shows.
enum MoneyFormat {

    /// The currency prefix D2' §A prints on every amount.
    ///
    /// **`Rs` is data, not copy** — three identical values in the three `Localizable.strings` files
    /// is exactly what `LocalizationTests` fails on, and a currency prefix is a proper noun. Same
    /// rule `+94` and the language endonyms follow.
    static let prefix = "Rs"

    /// What is drawn where an amount is not known **yet** — a quote in flight, or a fare that has not
    /// been computed (Δ C098).
    ///
    /// An em dash rather than `Rs 0`: a zero is a price, and a ride whose fare has not landed does not
    /// have one. `apps/passenger-android`'s `MoneyFormat.EMPTY` is the same character, and it is
    /// punctuation rather than copy — three identical values in the three `.strings` files is exactly
    /// what `LocalizationTests` reads as a translation nobody did.
    static let pending = "—"

    /// `74000` → `Rs 740`; `48050` → `Rs 480.50`. Whole rupees lose their `.00`.
    static func rupees(_ amountMinor: Int64) -> String {
        let negative = amountMinor < 0
        let absolute = negative ? -amountMinor : amountMinor
        let whole = absolute / minorUnits
        let cents = absolute % minorUnits

        let grouped = grouping.string(from: NSNumber(value: whole)) ?? String(whole)
        let text = cents == 0 ? grouped : grouped + "." + String(format: "%02lld", cents)
        return negative ? "\(prefix) -\(text)" : "\(prefix) \(text)"
    }

    /// The same, for a DTO that carries a `Money` rather than a bare `…Minor` field.
    static func rupees(_ money: Money) -> String { rupees(money.amountMinor) }

    /// `84` → `01:24` — SCR-PI-028's call timer (Δ C102).
    ///
    /// **Minutes and seconds, and they do not roll over into hours.** A call that ran past sixty
    /// minutes reads `61:07`, which is what the wireframe's two-field `01:24` means and what every
    /// telephone on the platform does; a third field would make a two-minute call read `00:02:14`.
    /// The same function is `apps/driver-ios`'s `MoneyFormat.timer(seconds:)` and
    /// `apps/passenger-android`'s `DateFormat.timer`.
    ///
    /// It lives here rather than in ``TripLabels`` because it is **not a clock**: there is no zone,
    /// no calendar and no locale in an elapsed duration, which is the whole reason D-38's Colombo
    /// rules do not apply to it.
    static func timer(seconds: Int64) -> String {
        let total = max(0, seconds)
        return String(format: "%02lld:%02lld", total / 60, total % 60)
    }

    private static let minorUnits: Int64 = 100

    private static let grouping: NumberFormatter = {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.groupingSeparator = ","
        formatter.usesGroupingSeparator = true
        return formatter
    }()
}
